pss/trojan_chunk.go

```diff
@@ -0,0 +1,221 @@
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package pss
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/binary"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/storage"
+)
+
+// MessageTopic is an alias for a 32 byte fixed-size array which contains an encoding of a message topic
+type MessageTopic [32]byte
+
+type trojanMessage struct {
+	length  [2]byte // big-endian encoding of message length
+	topic   MessageTopic
+	payload []byte
+	padding []byte
+}
+
+const trojanPayloadMaxSize = 4064                                // in bytes
+var trojanHashingFunc = storage.MakeHashFunc(storage.SHA3Hash)() // TOOD: make this work with storage.BMTHash
+
+// newMessageTopic creates a new MessageTopic variable with the given input string
+// the input string is taken as a byte slice and hashed
+func newMessageTopic(topic string) MessageTopic {
+	// TODO: is it ok to use this instead of `crypto.Keccak256`?
+	return MessageTopic(crypto.Keccak256Hash([]byte(topic)))
+}
+
+// newTrojanMessage creates a new trojanMessage variable with the given topic and message payload
+func newTrojanMessage(topic MessageTopic, payload []byte) (trojanMessage, error) {
+	if len(payload) > trojanPayloadMaxSize {
+		return trojanMessage{}, fmt.Errorf("trojan message payload cannot be greater than %d bytes", trojanPayloadMaxSize)
+	}
+
+	// get length as array of 2 bytes
+	payloadLength := uint16(len(payload))
+	lengthBuffer := make([]byte, 2)
+	binary.BigEndian.PutUint16(lengthBuffer, payloadLength)
+
+	// set random bytes as padding
+	paddingLength := trojanPayloadMaxSize - payloadLength
+	padding := make([]byte, paddingLength)
+	if _, err := rand.Read(padding); err != nil {
+		return trojanMessage{}, err
+	}
+
+	// create new trojan message var and set fields
+	tm := new(trojanMessage)
+	copy(tm.length[:], lengthBuffer[:])
+	tm.payload = payload
+	tm.padding = padding
+
+	return *tm, nil
+}
+
+// newTrojanChunk creates a new trojan chunk for the given targets and trojan message
+// TODO: discuss if instead of receiving a trojan message, we should receive a byte slice as payload
+func newTrojanChunk(targets [][]byte, message trojanMessage) (chunk.Chunk, error) {
+	if err := checkTargets(targets); err != nil {
+		return nil, err
+	}
+
+	// create span
+	span := newTrojanChunkSpan()
+
+	// find nonce for trojan chunk
+	nonce, addr, err := message.findNonce(span, targets)
+	if err != nil {
+		return nil, err
+	}
+
+	// serialize trojan message struct
+	m, err := message.MarshalBinary()
```

Comment:
these are done already within `findNonce`. wcich should just return the chunk and an error

Reply:
changed `iterateTrojanChunk` to return `(chunk.Chunk, err)`